spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/ribbon/okhttp/OkHttpStatusCodeException.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright 2013-2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.cloud.netflix.ribbon.okhttp;
+
+import okhttp3.Response;
+
+import java.io.IOException;
+import java.net.URI;
+import org.springframework.cloud.client.loadbalancer.RetryableStatusCodeException;
+
+/**
+ * An {@link RetryableStatusCodeException} that captures a {@link Response}
+ * @author Ryan Baxter
+ */
+public class OkHttpStatusCodeException extends RetryableStatusCodeException {
+	private Response response;
+
+	public OkHttpStatusCodeException(String serviceId, Response response, URI uri) throws IOException {
+		super(serviceId, response.code(), response, uri);
+		this.response = new Response.Builder().code(response.code()).message(response.message()).protocol(response.protocol())
+				.request(response.request()).headers(response.headers()).handshake(response.handshake())
+				.cacheResponse(response.cacheResponse()).networkResponse(response.networkResponse())
+				.priorResponse(response.priorResponse()).sentRequestAtMillis(response.sentRequestAtMillis())
+				.body(response.peekBody(Integer.MAX_VALUE)).build();
+		response.close();
```

Comment:
Moving this here seems very strange to me.

Reply:
Since the exception(s) were actually reading in the response, it seemed to make sense to me that it would be responsible for closing the response as well.  I suppose we can move it outside of the constructor, it just seemed disconnected that way.

Reply:
wrapping all this logic about copying over the response body in the exception also seemed to simplify the code in load balancing clients.  if we were to move the close call back to the load balancing clients than we would also have to move the code that copies the response body back there as well and it just added to something that was already complex.  since the fallback was using the response from the exception it just seemed to make the most sense to me to keep all that logic in the exception itself.